tools/codegen/gen_functionalization_type.py

```diff
@@ -0,0 +1,426 @@
+from tools.codegen.api import cpp
+from tools.codegen.api import dispatcher
+from tools.codegen.api.types import (
+    DispatcherSignature, CppSignatureGroup, Expr, Binding, NamedCType, ConstRefCType,
+    BaseCType, tensorT, intT, CType
+)
+from tools.codegen.api.translate import translate
+from tools.codegen.context import method_with_native_function
+from tools.codegen.model import (
+    Argument, NativeFunction, NativeFunctionsGroup, Return, SchemaKind, BackendIndex,
+    Variant, Tag, FunctionSchema, SelfArgument, TensorOptionsArguments, BaseType, BaseTy,
+    ListType
+)
+from typing import List, Optional, Union, Tuple, Sequence
+from dataclasses import dataclass
+from typing_extensions import Literal
+from tools.codegen.utils import mapMaybe, Target, assert_never
+
+from tools.autograd.gen_inplace_or_view_type import modifies_arguments
+
+def is_view_op(f: NativeFunction) -> bool:
+    rets = f.func.returns
+    is_non_mutating_view = len(rets) > 0 and any(r.annotation is not None and not r.annotation.is_write for r in rets)
+    is_inplace_view = f.tag is not None and f.tag is Tag.inplace_view
+    return is_non_mutating_view or is_inplace_view
```

Comment:
nit: this is probably worth canonizing on NativeFunction. It's too bad we can't detect this correctly on FunctionSchema